File: GPU Demo Mac/Process.m
//
//  Process.m
//  GPU Demo Mac
//
//  Created by Brent Gulanowski on 2018-08-14.
//  Copyright © 2018 Lichen Labs. All rights reserved.
//

#import "Process.h"

@interface Process ()
@property NSArray<id<GPUETask>> *tasks;
@end


@implementation Process

@synthesize engine = _engine;
@synthesize onComplete=_onComplete;

- (id<MTLDevice>)device {
    return _engine.device;
}

- (BOOL)needsRefresh {
    return NO;
}

- (instancetype)initWithEngine:(GPUEEngine *)engine host:(id<GPUERenderHost>)renderHost {
    self = [super init];
    if (self) {
        _engine = engine;
        id<GPUEBufferProducer> generate = [[GPUEComputePyramid alloc] initWithLibrary:engine.library size:1024u];
        id<GPUEBufferTransform> transform = [[GPUEBufferToTexture alloc] initWithLibrary:engine.library buffer:generate.outBuffer size:1024u];
        id<GPUETextureConsumer> draw = nil;
        
        if (renderHost) {
            draw = [[GPUEDrawTexture alloc] initWithLibrary:engine.library texture:transform.outTexture host:renderHost];
            self.tasks = @[generate, transform, draw];
        }
        else {
            self.tasks = @[ generate, transform];
        }
    }
    return self;
}

- (void)run {}

- (void)runWithDrawable:(id<MTLDrawable>)drawable {
    [_engine runTasks:self.tasks withDrawable:drawable];
}

- (instancetype)initWithEngine:(GPUEEngine *)engine {
    return [self initWithEngine:engine host:nil];
}

@end

File: GPU Demo iOS/Process.swift
//
//  Process.swift
//  GPU Demo iOS
//
//  Created by Brent Gulanowski on 2018-08-14.
//  Copyright © 2018 Lichen Labs. All rights reserved.
//

import UIKit
import GPUEngine

class Process: NSObject, GPUEProcess {
    
    let size : uint = 1024
    
    var device : MTLDevice {
        get {
            return engine.device
        }
    }
    
    lazy var tasks: [GPUETask] = {
        let size = vector_uint2(self.size, self.size)
        let gen = GPUEComputePyramid(library: engine.library, size: size)!
        let transform = GPUEBufferToTexture(library: engine.library, buffer: gen.outBuffer, size: size)!
        let draw = GPUEDrawTexture(library: engine.library, texture: transform.outTexture, host: renderHost)
        return [ gen, transform, draw ]
    }()
    
    var needsRefresh: Bool {
        get {
            return true
        }
    }
    
    var refresh: (() -> Void)!
    
    var onComplete: (() -> Void)!
    
    func run() {
        engine.run(tasks)
    }
    
    func run(with drawable: MTLDrawable!) {
        engine.run(tasks, with: drawable)
    }
    
    var engine: GPUEEngine!
    var renderHost : GPUERenderHost!

    required init!(engine: GPUEEngine!) {
        self.engine = engine
    }
    
    init(engine: GPUEEngine, renderHost: GPUERenderHost) {
        self.engine = engine
        self.renderHost = renderHost
    }
}
